core/reactor_common.c

```diff
@@ -1316,18 +1316,70 @@ bool _lf_check_deadline(self_base_t* self, bool invoke_deadline_handler) {
     return false;
 }
 
+#ifdef LF_THREADED
+
+void* run_watchdog(watchdog_t* watchdog) {
+    watchdog->thread_active = true;
+    self_base_t* base = watchdog->base;
+
+    while (lf_time_physical() < watchdog->expiration) {
+        interval_t T = watchdog->expiration - lf_time_physical();
+        lf_mutex_unlock(&(base->watchdog_mutex));
+        lf_nanosleep(T);
+        lf_mutex_lock(&(base->watchdog_mutex));
+    }
+
+
+    watchdog_function_t watchdog_func = watchdog->watchdog_function;
+    (*watchdog_func)(watchdog);
+    lf_mutex_unlock(&(base->watchdog_mutex));
```

Comment:
When you return here, the thread_active field is still true. So there wont be another thread created. Also, shouldnt the thread keep looping inside this function and wait for a watchdog reset?

Reply:
I am pretty sure we wouldnt want that. Creating a thread has a significant overhead. It would be better to have a single thread per watchdog. But you do want it to do an actual sleep while it is waiting, not a busy-sleep